### PR TITLE
Allow context argument to be nil  

### PR DESCRIPTION
**Pull request by bakape**
Allows passing a nil `context.Context` to all external API endpoints. As a result, you can now write
```go
    conn.Exec(nil, `select 1`)
```
instead of
```go
    conn.Exec(context.Background(), `select 1`)
```
.

Additionally, nil contexts now avoid any context synchronization code and thus result in a 3-10% performance increase in the benchmarks. Note, that sometimes there will be spikes due to calling Postgres - an external process doing its own thing - but nil context cases are faster on average. 

I have covered this in all unit tests that might now or in the future cause a nil pointer deref due to this addition. 

Will  also make PRs for the pgx and pgxpool packages at a later date. Simply started from the bottom.

### CODE REVIEW

**Comment by jackc** [comment]
Thanks for developing this optimization and especially including tests.

I ran the benchmarks on my machine.

```
jack@glados ~/dev/pgconn ±bakape-master » go test -benchtime=15s -bench=. -benchmem
Alias tip: got -benchtime=15s -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/jackc/pgconn
BenchmarkConnect/Unix_socket-16             	    5570	   3222607 ns/op	   22262 B/op	     140 allocs/op
BenchmarkConnect/TCP-16                     	    1555	  11467678 ns/op	  112286 B/op	     918 allocs/op
BenchmarkConnect/Unix_socket_nil_context-16 	    5497	   3256045 ns/op	   22259 B/op	     140 allocs/op
BenchmarkConnect/TCP_nil_context-16         	    1560	  11458122 ns/op	  112277 B/op	     918 allocs/op
BenchmarkExec/background_context-16         	  565028	     31454 ns/op	     126 B/op	       0 allocs/op
BenchmarkExec/nil_context-16                	  555786	     31167 ns/op	     126 B/op	       0 allocs/op
BenchmarkExecPossibleToCancel-16            	  515037	     34404 ns/op	     126 B/op	       0 allocs/op
BenchmarkExecPrepared/background_context-16 	  812840	     22099 ns/op	     131 B/op	       0 allocs/op
BenchmarkExecPrepared/nil_context-16        	  832257	     21360 ns/op	     131 B/op	       0 allocs/op
BenchmarkExecPreparedPossibleToCancel-16    	  685942	     25794 ns/op	     131 B/op	       0 allocs/op
PASS
ok  	github.com/jackc/pgconn	185.122s
```

`Exec` had a 0.9% improvement and `ExecPrepared` had a 3.5% improvement with a `nil` context vs `context.Background()`. That's an improvement, but not as much as you saw. Perhaps it is somewhat platform specific?

One thing that concerns me is that this is essentially encouraging people to go directly  against the context documentation. I'm not comfortable with that. From https://godoc.org/context

> Programs that use Contexts should follow these rules to keep interfaces consistent across packages and enable static analysis tools to check context propagation:

> Do not pass a nil Context, even if a function permits it.

Now that said, even the standard library violates the context rules by storing a context in the http.Request struct and HTTP routers tend to bend the rules on storing optional arguments in context. So I guess this isn't necessarily a blocker, but it does give me pause...

Would we get the same performance benefit by testing `if ctx != context.Background()` instead of testing `if ctx != nil`?  If yes, that would be ideal.

**Comment by bakape** [comment]
> `Exec` had a 0.9% improvement and `ExecPrepared` had a 3.5% improvement with a `nil` context vs `context.Background()`. That's an improvement, but not as much as you saw. Perhaps it is somewhat platform specific?

Perhaps it is. Numbers also seem to vary from run to run, so best run it a few times. Mine seem to hover between 3-10%. 10% is likely just random runtime spikes, so 3-5% should be more accurate.

>>Programs that use Contexts should follow these rules to keep interfaces consistent across packages and enable static analysis tools to check context propagation:
Do not pass a nil Context, even if a function permits it.

>Now that said, even the standard library violates the context rules by storing a context in the http.Request struct and HTTP routers tend to bend the rules on storing optional arguments in context. So I guess this isn't necessarily a blocker, but it does give me pause...

I'd say it's just an unintended deoptimization locked into the stdlib by the Go1 guarantee, same as map itteration order and the Marshaller interfaces. Also, a "good style" rule. "Good style" is often more verbose and slower.  After all, if the standard library was perfect, we'd all be using `db/sql`.

>Would we get the same performance benefit by testing if ctx != context.Background() instead of testing if ctx != nil? If yes, that would be ideal.

Yes, best check for both. That way we can have the cleaner option of passing `nil` and allow existing code using `context.Background()` to benefit as well.

**Comment by bakape** [comment]
Oh, I understand now, why the numbers are so random. What this optimization does is create a branch in the compiled code, where it is possible to avoid a few context switches to other goroutines. Because `context.Background()` returns a pointer to `context.background` as an interface type, 
```go
func Background() Context {
	return background
}
```
 will never be inlined. Because of that, the compiler decides a context switch might occur there. So, as with any channel receive, the compiler creates a stop point for the goroutine on `<-ctx.Done()`. The runtime then possibly switches to a different goroutine at that point. Of course, context switches aren't free and the runtime will have to switch to the previous goroutine at some point or even copy its context to run on another thread.  

tl;dr: Performance differs, if the runtime decides to switch to another goroutine on `<-ctx.Done()`. We can avoid that.

**Comment by jackc** [comment]
So if I'm reading this correctly, we now get the same optimization with `context.Background()` as with `nil`.

If that is the case, I would strongly prefer _not_ to accept `nil` at all. i.e. I think it is better to consistently crash if a `nil` context is received.

1. This follows the Go doc / style guides for context.
2. If any new code omits this change, it is only slightly slower instead of crashing on something we document as acceptable.
3. But most importantly, sometimes the context is passed back to user supplied code (e.g. the pgx Logger interface). This could lead to edge cases that could easily slip through testing. e.g. the production logger crashes with nil context, but a different logger or no logger at all is used in testing.

**Comment by bakape** [comment]
>So if I'm reading this correctly, we now get the same optimization with
context.Background() as with nil.
Correct.

>If that is the case, I would strongly prefer *not* to accept nil at all.
i.e. I think it is better to consistently crash if a nil context is
received.
>This follows the Go doc / style guides for context.
>If any new code omits this change, it is only slightly slower instead of
crashing on something we document as acceptable.
>But most importantly, sometimes the context is passed back to user
supplied code (e.g. the pgx Logger interface). This could lead to edge
cases that could easily slip through testing. e.g. the production logger
crashes with nil context, but a different logger or no logger at all is
used in testing.
I understand your reasoning but you are trading API ease of use for no
immediate benefit and only possible future benefit. The standard library
does not have to make this tradeoff, as it has functions that do not accept
a context. Those call their own contextual versions with
`context.Background()`. If you consider it a better alternative, we could
do something akin and each public function, that accepts a context, default
to `context.Background()` via a null check. Spamming `context.Background()`
in most of your function calls, when you could be using a `nil` is in no
way a clen API. Why can there be nil pointers to option structs but not nil
contexts?

**Comment by jackc** [comment]
> I understand your reasoning but you are trading API ease of use for no
immediate benefit and only possible future benefit.

The immediate benefit is the reduced complexity / surface area. We are essentially enforcing a not null constraint on `ctx`.

> The standard library
does not have to make this tradeoff, as it has functions that do not accept
a context. Those call their own contextual versions with
`context.Background()`.

This is true, but I don't think it was for this reason. Backwards compatibility made it impossible to change the non-context functions. So new functions had to be added. I believe that any new functions would only come in context versions. See this comment by the primary maintainer of `database/sql` for an example https://github.com/golang/go/issues/22697#issuecomment-344095510 -- he would remove the non-context functions if he could.

> Spamming `context.Background()`
in most of your function calls, when you could be using a `nil` is in no
way a clen API. Why can there be nil pointers to option structs but not nil
contexts?

I agree that `context.Background()` is verbose, but I would suggest that production code should only rarely use `context.Background()`. Otherwise, any DB call could potentially hang indefinitely. For web servers, at least the request context should be used, often with an additional timeout.

> If you consider it a better alternative, we could
do something akin and each public function, that accepts a context, default
to `context.Background()` via a null check. 

I don't think that would be appropriate in `pgconn`; the vast majority of users should never be directly accessing `pgconn`. It would make more sense in `pgx`, but even there `context.Background()` usage is not something I want to encourage. I think a 3rd party wrapper package would make the more sense. I think it would only be 50-100 lines of code.

**Comment by bakape** [comment]
Removed everything except `context.Background()` optimization and benchmarks. Performance increase still randomly fluctuates between 1-10% due to runtime goroutine management choices on my machine.

**Comment by jackc** [comment]
Thanks!
